### PR TITLE
Persist Token account indexes via plugin

### PR DESCRIPTION
**Pull request by lijunwangs**
Persisting token owner index and token mint index if configured so.
Do bulk insert during startup
Persist during accounts update due to transaction.

### CODE REVIEW

**Comment by carllin** [comment]
@lijunwangs is there a guarantee/verification that these indexes being persisted are consistent up to some snapshot/root `X`? 

For example what happens if:

1. a node crashes while writing to the index
2. the validator starts from a different snapshot in the past/future

**Comment by lijunwangs** [comment]
> 



> @lijunwangs what is the use case for persisting the indexes, faster startup?
> 
> is there a guarantee/verification that these indexes being persisted are consistent up to some snapshot/root `X`?
> 
> For example what happens if:
> 
> 1. a node crashes while writing to the index
> 2. the validator starts from a different snapshot in the past/future

I was thinking the plugin could choose to blow away the index at restart in case of crash. Alternatively we can store the slot info in the index similar to the accounts data.

**Comment by carllin** [comment]
> I was thinking the plugin could choose to blow away the index at restart in case of crash.

Ok on the same page then, we should blow away the index on every restart and rebuild the index.

**Comment by lijunwangs** [comment]
> > I was thinking the plugin could choose to blow away the index at restart in case of crash.
> 
> Ok on the same page then, we should blow away the index on every restart and rebuild the index.

On second thought, I think adding the slot info in the index makes it more flexible.

File: src/accountsdb_plugin_postgres.rs [hunk 1]
```diff
@@ -32,21 +32,55 @@ impl std::fmt::Debug for AccountsDbPluginPostgres {
...
+    /// Specify the path to PostgreSQL server's certificate file
     pub server_ca: Option<String>,
+
+    /// Specify the path to the local client's certificate file
     pub client_cert: Option<String>,
+
+    /// Specify the path to the local client's private PEM key file.
     pub client_key: Option<String>,
+
+    /// Controls if to index the token owners. The default is false
```
Comment (carllin) [inline src/accountsdb_plugin_postgres.rs:79]: nit: if -> whether

File: src/accountsdb_plugin_postgres.rs [hunk 2]
```diff
@@ -32,21 +32,55 @@ impl std::fmt::Debug for AccountsDbPluginPostgres {
...
+    pub index_token_owner: Option<bool>,
+
+    /// Controls if to index the token mints. The default is false
```
Comment (carllin) [inline src/accountsdb_plugin_postgres.rs:82]: nit: if -> whether

File: src/inline_spl_token.rs [hunk 3]
```diff
@@ -0,0 +1,92 @@
+/// Partial SPL Token declarations inlined to avoid an external dependency on the spl-token crate
+/// Copied from solana-runtime
```
Comment (carllin) [inline src/inline_spl_token.rs:2]: Adding the dependency seems like the right call here

File: src/inline_spl_token_2022.rs [hunk 4]
```diff
@@ -0,0 +1,19 @@
+/// Partial SPL Token declarations inlined to avoid an external dependency on the spl-token-2022 crate
+/// Copied from solana-runtime
```
Comment (carllin) [inline src/inline_spl_token_2022.rs:2]: Adding the dependency seems like the right call here

File: src/postgres_client/postgres_client_account_index.rs [hunk 5]
```diff
@@ -0,0 +1,419 @@
...
+                self.batch_size,
+                10000,
+                10000
+            );
+        }
+        Ok(())
+    }
+
+    /// Generic function to queue the token owner index for bulk insert.
+    fn queue_token_owner_index_gen<G: GenericTokenAccount>(
```
Comment (carllin) [inline src/postgres_client/postgres_client_account_index.rs:254]: nit: not too sure what the `_gen` suffix stands for, seems like we can just remove it

File: src/postgres_client/postgres_client_account_index.rs [hunk 6]
```diff
@@ -0,0 +1,419 @@
...
+            );
+            self.queue_token_mint_index_gen::<inline_spl_token_2022::Account>(
+                &inline_spl_token_2022::id(),
+                account,
+            );
+        }
+    }
+
+    /// Generic function to update a single token owner index.
+    fn update_token_owner_index_gen<G: GenericTokenAccount>(
```
Comment (carllin) [inline src/postgres_client/postgres_client_account_index.rs:317]: nit: not too sure what the `_gen` suffix stands for, seems like we can just remove it

File: src/postgres_client/postgres_client_account_index.rs (hunk 5 above)
Comment (carllin) [inline src/postgres_client/postgres_client_account_index.rs:254]: nit: spell out gen-> generic

File: src/postgres_client/postgres_client_account_index.rs [hunk 7]
```diff
@@ -0,0 +1,419 @@
...
+                    owner: owner_key,
+                    inner_key: pubkey.to_vec(),
+                    slot: account.slot,
+                });
+            }
+        }
+    }
+
+    /// Generic function to queue the token mint index for bulk insert.
+    fn queue_token_mint_index_gen<G: GenericTokenAccount>(
```
Comment (carllin) [inline src/postgres_client/postgres_client_account_index.rs:273]: nit: spell out gen-> generic

File: src/postgres_client/postgres_client_account_index.rs [hunk 8]
```diff
@@ -0,0 +1,419 @@
...
+                    return Err(AccountsDbPluginError::AccountsUpdateError { msg });
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Generic function to update a single token mint index.
+    fn update_token_mint_index_gen<G: GenericTokenAccount>(
```
Comment (carllin) [inline src/postgres_client/postgres_client_account_index.rs:344]: nit: spell out gen-> generic

**Comment by carllin** [comment]
> On second thought, I think adding the slot info in the index makes it more flexible.

Yeah if we can add the slot info to the secondary index lookups, it should make the lookups even faster than they are today. Today, we get a blob of account keys  and then have to do another lookup in the primary index to find the slots

File: src/postgres_client.rs [hunk 9]
```diff
@@ -52,11 +56,19 @@ struct PostgresSqlClientWrapper {
...
 }
 
 pub struct SimplePostgresClient {
     batch_size: usize,
     pending_account_updates: Vec<DbAccountInfo>,
+    index_token_owner: bool,
+    index_token_mint: bool,
+    pending_token_owner_index: Vec<TokenSecondaryIndex>,
+    pending_token_mint_index: Vec<TokenSecondaryIndex>,
     client: Mutex<PostgresSqlClientWrapper>,
```
Comment (carllin) [inline src/postgres_client.rs:72]: It seems like the inner client is the only thing that needs to be in the mutex, the statements can be extracted outside so we don't need to grab a lock to access the tatements

File: src/postgres_client/postgres_client_account_index.rs [hunk 10]
```diff
@@ -0,0 +1,419 @@
...
+            );
+            inc_new_counter_debug!(
+                "accountsdb-plugin-postgres-update-mint-index-count",
+                self.batch_size,
+                10000,
+                10000
+            );
+        }
+        Ok(())
+    }
```
Comment (carllin) [inline src/postgres_client/postgres_client_account_index.rs:251]: Can we condense these two functions in to a generic function that takes the:

1. index: `Vec<TokenSecondaryIndex>`
2. prefix string for the metrics
3. Once this is implemented: https://github.com/solana-labs/solana-accountsdb-plugin-postgres/pull/9/files#r815277811, pass the statement you need in
4.

File: src/postgres_client/postgres_client_account_index.rs [hunk 11]
```diff
@@ -0,0 +1,419 @@
...
+                return Err(AccountsDbPluginError::Custom(Box::new(AccountsDbPluginPostgresError::DataSchemaError {
+                    msg: format!(
+                        "Error in preparing for the token mint index update PostgreSQL database: {} host: {:?} user: {:?} config: {:?}",
+                        err, config.host, config.user, config
+                    ),
+                })));
+            }
+            Ok(update_account_stmt) => Ok(update_account_stmt),
+        }
+    }
```
Comment (carllin) [inline src/postgres_client/postgres_client_account_index.rs:135]: Same as https://github.com/solana-labs/solana-accountsdb-plugin-postgres/pull/9/files#r815277859, I believe these can be condensed together

File: src/postgres_client.rs (hunk 9 above)
Comment (lijunwangs) [reply src/postgres_client.rs:72]: The prepared statements are created with a certain client which is used during prepare-- we cannot use a statement out side of the client creating it. So I think it makes sense putting them together.

File: src/postgres_client/postgres_client_account_index.rs (hunk 8 above)
Comment (lijunwangs) [reply src/postgres_client/postgres_client_account_index.rs]: done

File: src/postgres_client/postgres_client_account_index.rs (hunk 7 above)
Comment (lijunwangs) [reply src/postgres_client/postgres_client_account_index.rs]: done

File: src/inline_spl_token.rs (hunk 3 above)
Comment (lijunwangs) [reply src/inline_spl_token.rs:2]: Yes. I have created a PR to make this module public in solana-runtime. Once that is published and available I will remove from this

File: scripts/create_schema.sql [hunk 12]
```diff
@@ -177,18 +177,22 @@ CREATE TABLE block (
...
 );
 
 CREATE INDEX spl_token_owner_index_owner_key ON spl_token_owner_index (owner_key);
+CREATE UNIQUE INDEX spl_token_owner_index_owner_pair ON spl_token_owner_index (owner_key, inner_key);
 
 -- The table storing spl mint to account indexes
 CREATE TABLE spl_token_mint_index (
     mint_key BYTEA NOT NULL,
-    inner_key BYTEA NOT NULL
+    inner_key BYTEA NOT NULL,
```
[the remark this answers is not in the document]
Comment (lijunwangs) [reply scripts/create_schema.sql]: done

File: src/accountsdb_plugin_postgres.rs (hunk 2 above)
Comment (lijunwangs) [reply src/accountsdb_plugin_postgres.rs]: done

File: src/inline_spl_token.rs [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+        mint: Pubkey,
+        owner: Pubkey,
+        amount: u64,
+        delegate: COption<Pubkey>,
+        state: AccountState,
+        is_native: COption<u64>,
+        delegated_amount: u64,
+        close_authority: COption<Pubkey>,
+    }
+*/
```
[the remark this answers is not in the document]
Comment (lijunwangs) [reply src/inline_spl_token.rs]: removed

File: src/postgres_client/postgres_client_account_index.rs (hunk 6 above)
Comment (lijunwangs) [reply src/postgres_client/postgres_client_account_index.rs]: done

File: src/postgres_client/postgres_client_account_index.rs (hunk 5 above)
Comment (lijunwangs) [reply src/postgres_client/postgres_client_account_index.rs]: done

File: src/postgres_client/postgres_client_account_index.rs (hunk 10 above)
Comment (lijunwangs) [reply src/postgres_client/postgres_client_account_index.rs]: done

File: src/postgres_client/postgres_client_account_index.rs (hunk 11 above)
Comment (lijunwangs) [reply src/postgres_client/postgres_client_account_index.rs]: done

File: src/inline_spl_token_2022.rs (hunk 4 above)
Comment (lijunwangs) [reply src/inline_spl_token_2022.rs:2]: @carllin -- as you might be aware that from Trent we intentionally keep these modules internal. The alternative is use spl-token which will requires us to unpack the whole spl-token account -- which is probably unnecessarily more expensive as I am only interested in the mint and owner field. Can I get a pass to merge this as is to unblock? Thanks!

Comment (mvines) [reply src/inline_spl_token_2022.rs:2]: How about working upstream to put this kind of partial-unpacking functionally into spl-token, so you can import that crate from here. This self-inflicted duplication seems quite avoidable. @joncinque's a good resource for guidance on the spl-token crate side

File: src/accountsdb_plugin_postgres.rs [hunk 14]
```diff
@@ -32,21 +32,55 @@ impl std::fmt::Debug for AccountsDbPluginPostgres {
     }
 }
 
+/// The Configurqation for the PostgreSQL plugin
```
Comment (carllin) [inline src/accountsdb_plugin_postgres.rs:35]: nit: Configurqation -> Configuration

File: src/postgres_client/postgres_client_account_index.rs [hunk 15]
```diff
@@ -0,0 +1,377 @@
...
+        WHERE owner_index.slot < excluded.slot";
+
+        Self::prepare_query_statement(client, config, stmt)
+    }
+
+    pub fn build_single_token_mint_index_upsert_statement(
+        client: &mut Client,
+        config: &AccountsDbPluginPostgresConfig,
+    ) -> Result<Statement, AccountsDbPluginError> {
+        let stmt = "INSERT INTO spl_token_mint_index AS mint_index (mint_key, account_key, slot) \
```
Comment (carllin) [inline src/postgres_client/postgres_client_account_index.rs:49]: can we make these const strings, i.e. `const BULK_MINT_INDEX_INSERT`: &str = "";

File: src/postgres_client/postgres_client_account_index.rs [hunk 16]
```diff
@@ -0,0 +1,377 @@
...
+    pub fn build_single_token_owner_index_upsert_statement(
+        client: &mut Client,
+        config: &AccountsDbPluginPostgresConfig,
+    ) -> Result<Statement, AccountsDbPluginError> {
+        let stmt =
+            "INSERT INTO spl_token_owner_index AS owner_index (owner_key, account_key, slot) \
+        VALUES ($1, $2, $3) \
+        ON CONFLICT (owner_key, account_key) \
+        DO UPDATE SET slot=excluded.slot \
+        WHERE owner_index.slot < excluded.slot";
```
Comment (carllin) [inline src/postgres_client/postgres_client_account_index.rs:40]: can we make these const strings, i.e. const BULK_TOKEN_OWNER_INDEX_INSERT_STATEMENT: &str = "";

File: src/postgres_client/postgres_client_account_index.rs [hunk 17]
```diff
@@ -0,0 +1,377 @@
...
+    postgres::{Client, Statement},
+    solana_accountsdb_plugin_interface::accountsdb_plugin_interface::AccountsDbPluginError,
+    solana_measure::measure::Measure,
+    solana_metrics::*,
+    solana_sdk::pubkey::Pubkey,
+    tokio_postgres::types,
+};
+
+const TOKEN_INDEX_COLUMN_COUNT: usize = 3;
+/// Model the secondary index for token owner and mint
```
Comment (carllin) [inline src/postgres_client/postgres_client_account_index.rs:23]: nit: This comment currently isn't very helpful, maybe an explanation of:
1. How this secondary index works
2. How each of the fields is necessary for this index

File: src/postgres_client/postgres_client_account_index.rs [hunk 18]
```diff
@@ -0,0 +1,377 @@
...
+            "mint_key",
+            config,
+        )
+    }
+
+    /// Execute the common token bulk insert query.
+    fn bulk_insert_token_index_common(
+        batch_size: usize,
+        client: &mut Client,
+        indexes: &mut Vec<TokenSecondaryIndex>,
```
Comment (carllin) [inline src/postgres_client/postgres_client_account_index.rs:135]: nit: 1indexes -> index_entries`

File: src/postgres_client/postgres_client_account_index.rs [hunk 19]
```diff
@@ -0,0 +1,377 @@
...
+    solana_accountsdb_plugin_interface::accountsdb_plugin_interface::AccountsDbPluginError,
+    solana_measure::measure::Measure,
+    solana_metrics::*,
+    solana_sdk::pubkey::Pubkey,
+    tokio_postgres::types,
+};
+
+const TOKEN_INDEX_COLUMN_COUNT: usize = 3;
+/// Model the secondary index for token owner and mint
+pub struct TokenSecondaryIndex {
```
Comment (carllin) [inline src/postgres_client/postgres_client_account_index.rs:24]: nit: `TokenSecondaryIndex -> TokenSecondaryIndexEntry`

**Review by carllin — APPROVED** [review]
Approved with some nits

File: scripts/postgresql.conf [hunk 20]
```diff
@@ -411,7 +411,7 @@ max_wal_senders = 0		# max number of walsender processes
 #geqo_threshold = 12
 #geqo_effort = 5			# range 1-10
 #geqo_pool_size = 0			# selects default based on effort
-#geqo_generations = 0			# selects default based on effort
+#geqo_genericerations = 0			# selects default based on effort
```
Comment (joncinque) [inline scripts/postgresql.conf:414]: Looks like this got blasted by a global search/replace

File: scripts/postgresql.conf [hunk 21]
```diff
@@ -424,7 +424,7 @@ max_wal_senders = 0		# max number of walsender processes
 #jit = on				# allow JIT compilation
 #join_collapse_limit = 8		# 1 disables collapsing of explicit
 					# JOIN clauses
-#plan_cache_mode = auto			# auto, force_generic_plan or
+#plan_cache_mode = auto			# auto, force_genericeric_plan or
```
Comment (joncinque) [inline scripts/postgresql.conf:427]: Same here

File: src/postgres_client/postgres_client_account_index.rs [hunk 22]
```diff
@@ -0,0 +1,377 @@
...
+    }
+
+    /// Generic function to update a single token mint index.
+    fn update_token_mint_index_generic<G: GenericTokenAccount>(
+        client: &mut Client,
+        statement: &Statement,
+        token_id: &Pubkey,
+        account: &DbAccountInfo,
+    ) -> Result<(), AccountsDbPluginError> {
+        if account.owner() == token_id.to_bytes() {
```
Comment (joncinque) [inline src/postgres_client/postgres_client_account_index.rs:308]: nittiest of nits: I imagine this is a hot path, so you can potentially speed this up by replacing `to_bytes()` with `as_ref()` everywhere

**Review by joncinque — COMMENTED** [review]
Just a few little additional comments -- once we can get your spl-token crate out, I'm happy to help out getting the inline bits removed.

File: src/accountsdb_plugin_postgres.rs (hunk 14 above)
Comment (lijunwangs) [reply src/accountsdb_plugin_postgres.rs]: Done

File: src/postgres_client/postgres_client_account_index.rs (hunk 15 above)
Comment (lijunwangs) [reply src/postgres_client/postgres_client_account_index.rs]: Done

File: src/postgres_client/postgres_client_account_index.rs (hunk 16 above)
Comment (lijunwangs) [reply src/postgres_client/postgres_client_account_index.rs]: Done

File: src/postgres_client/postgres_client_account_index.rs (hunk 17 above)
Comment (lijunwangs) [reply src/postgres_client/postgres_client_account_index.rs]: Updated

File: src/postgres_client/postgres_client_account_index.rs [hunk 23]
```diff
@@ -0,0 +1,377 @@
...
+    solana_measure::measure::Measure,
+    solana_metrics::*,
+    solana_sdk::pubkey::Pubkey,
+    tokio_postgres::types,
+};
+
+const TOKEN_INDEX_COLUMN_COUNT: usize = 3;
+/// Model the secondary index for token owner and mint
+pub struct TokenSecondaryIndex {
+    owner: Vec<u8>,
```
[the remark this answers is not in the document]
Comment (lijunwangs) [reply src/postgres_client/postgres_client_account_index.rs]: Done

File: src/postgres_client/postgres_client_account_index.rs (hunk 18 above)
Comment (lijunwangs) [reply src/postgres_client/postgres_client_account_index.rs]: Done

File: src/postgres_client/postgres_client_account_index.rs (hunk 19 above)
Comment (lijunwangs) [reply src/postgres_client/postgres_client_account_index.rs]: Done

File: scripts/postgresql.conf (hunk 20 above)
Comment (lijunwangs) [reply scripts/postgresql.conf]: Good catch, fixed

File: scripts/postgresql.conf (hunk 21 above)
Comment (lijunwangs) [reply scripts/postgresql.conf]: Fixed

File: src/postgres_client/postgres_client_account_index.rs (hunk 22 above)
Comment (lijunwangs) [reply src/postgres_client/postgres_client_account_index.rs]: Great! Thanks

**Pull request merged**